packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/assign-public-ip/lambda/lib/route53.py

```diff
@@ -0,0 +1,91 @@
+from dataclasses import dataclass
+from typing import *
+import logging
+
+
+@dataclass
+class Route53RecordSetLocator:
+    hosted_zone_id: str
+    record_name: str
+
+
+# TODO: NEEDS OVERHAUL
+class Route53RecordSetMerger:
+    route53_client: Any
+    dynamo_table_client: Any
+
+    def __init__(self, route53_client):
+        self.route53_client = route53_client
+
+    def merge(self, locator: Route53RecordSetLocator, adds: List[str] = None, removes: List[str] = None):
+        adds = [] if adds is None else adds
+        removes = [] if removes is None else removes
+
+        record_set, is_new = self.get_record_set(locator)
+        new_record_set = merge_record_set_changes(record_set=record_set, adds=adds, removes=removes)
+
+        if len(new_record_set['ResourceRecords']) > 0:
+            self.upsert(locator, new_record_set)
+        elif not is_new:
+            self.delete(locator, record_set)
+        else:
+            logging.warning('Refusing to do anything with an new but empty recordset')
+
+    def get_record_set(self, locator: Route53RecordSetLocator):
+        record_type = 'A'
+        result = self.route53_client.list_resource_record_sets(HostedZoneId=locator.hosted_zone_id,
+                                                               StartRecordName=locator.record_name,
+                                                               StartRecordType=record_type, MaxItems="1")
+
+        record_sets = result['ResourceRecordSets']
+        if len(record_sets) > 0:
+            logging.info(f'Found a pre-existing record set: {record_sets[0]}')
+            return record_sets[0], False
+
+        logging.info('Creating a new record set')
+        return {'Name': locator.record_name, 'Type': record_type, 'ResourceRecords': [], 'TTL': 60}, True
+
+    def upsert(self, locator: Route53RecordSetLocator, record_set):
+        logging.info(f'Upserting record set {record_set}')
+        self.route53_client.change_resource_record_sets(
+            HostedZoneId=locator.hosted_zone_id, ChangeBatch={
+                'Comment': 'Automatic',
+                'Changes': [{
+                    'Action': 'UPSERT',
+                    'ResourceRecordSet': record_set
+                }]
+            })
+
+    def delete(self, locator: Route53RecordSetLocator, record_set):
+        logging.info(f'Deleting record set: {record_set}')
+        self.route53_client.change_resource_record_sets(
```

Comment:
Same here, maybe a helper method with exponential backoff on the retry so that way it can handle the 5/sec default quota on Route53 API calls as well:

> Five requests per second per AWS account. If you submit more than five requests per second, Amazon Route 53 returns an HTTP 400 error (Bad request). The response header also includes a Code element with a value of Throttling and a Message element with a value of Rate exceeded.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-requests

Reply:
I've added a catch for this in a recent commit.